pkg/fake_pub_server/lib/fake_pub_server.dart

```diff
@@ -84,3 +86,21 @@ class FakePubServer {
     _logger.info('closed');
   }
 }
+
+// Normally this method could be just void, however reading the shelf.Response's
+// body forces us to return a new instance of the response.
+Future<shelf.Response> _validateAndWrapResponse(
+    shelf.Request rq, shelf.Response rs) async {
+  if (rs == null) return null;
+
+  final ct = rs.headers[HttpHeaders.contentTypeHeader];
+  if (ct == null || ct.isEmpty) {
+    // TODO: force headers everywhere and throw exception instead of logging
+    _logger.warning('Content type header is missing for ${rq.requestedUri}.');
+  } else if (ct.contains('text/html')) {
+    final body = await rs.readAsString();
+    parseAndValidateHtml(body);
```

Comment:
Okay, I see that doing it here ensures that it always happens, because we're doing it on the return-path from the server.

But it seems more natural to put it in the test code that is actually calling it... that way it also gets run when we do integration tests against production? Another upside with moving this check into the integration test, is that we have less customization of the fake server, which means it's closer to the real thing, and more likely to produce the same bugs.

Another option is to put it on the server, whenever we create an HTML response.. and then have it be present as an assertion... It's probably too expensive to do that :D   Then one could do an assertion that only runs when we're configured in test-mode, but that's probably how we got to this..

I could be sold on either solution, but keeping the fake server as simple as possible might be worth considering.

Reply:
Hm, we could move these to the integration test's client side, if we (a) check the HTML structure as the http client returns it, and (b) intercept it in puppeteer and do the same check there too.

The tricky part is where to share the code. `pkg/pub_integration` does not depend on `app/` yet, and I think it has benefits to keep it that way (it handles it as black box, while fake_pub_server actively uses pub code). Maybe `pkg/shared_validation`?

Reply:
Discussed offline. I'll merge this and will move it to a new package.